Tighten kubectl version check

caesarxuchao opened the pull request:
Fix https://github.com/kubernetes/kubernetes/issues/40813

This let api server apply the workaround added in https://github.com/kubernetes/kubernetes/pull/35840 only for kubectl version >=1.3.0 and <1.5.0.

~~Note that if a kubectl is built from a commit between 1.3.0 and 1.50 in the master branch, kubectl will have version `kubectl/v0.0.0`, so the workaround will not be applied and such kubectl will suffer https://github.com/kubernetes/kubernetes/issues/35791. I think that's a less evil bug than https://github.com/kubernetes/kubernetes/issues/40813, thus we should merge this pull.~~

Commit:
tighten kubectl version check

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42600)
<!-- Reviewable:end -->

foxish commented:
/lgtm

liggitt reviewed (commented):
why wouldn't we tightly bound the versions the workaround applies to? (>= 1.4.0, <1.5.0)

Are we also fixing the v0.0.0 issue in head?

caesarxuchao commented:
> why wouldn't we tightly bound the versions the workaround applies to? (>= 1.4.0, <1.5.0)

Thanks, i updated the code as suggested. Also updated the PR description. PTAL.

> Are we also fixing the v0.0.0 issue in head?

i think it has always been like this. Only when we make a release will the version be written to [base.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/version/base.go), so it's not a bug.

liggitt commented:
>> Are we also fixing the v0.0.0 issue in head?
>
> i think it has always been like this. Only when we make a release will the version be written to base.go, so it's not a bug.

I don't think it has. Building from a branch off of master just before the release-1.5 branch was cut (a3ba760) with a random commit of my own added on top, I get

```
User-Agent: kubectl/v1.5.0 (darwin/amd64) kubernetes/a3ba760
```

the issue might be that the build scripts only set vars under `k8s.io/kubernetes/pkg/version`, and once we switched to the client-go base.go, it stopped getting dynamic vars
```
kube::version::ldflag() {
  local key=${1}
  local val=${2}

  echo "-X ${KUBE_GO_PACKAGE}/pkg/version.${key}=${val}"
}
```

liggitt commented on pkg/apiserver/apiserver.go:
```diff
@@ -346,8 +346,8 @@ func APIVersionHandler(s runtime.NegotiatedSerializer, getAPIVersionsFunc func(r
-func isOldKubectl(userAgent string) bool {
+// TODO: Remove in 1.6. Returns if kubectl is older than v1.5.0 and newer than or equal to v1.4.0
+func is14Kubectl(userAgent string) bool {
 	// example userAgent string: kubectl-1.3/v1.3.8 (linux/amd64) kubernetes/e328d5b
```
funnily, the example user agent string no longer passes this function

caesarxuchao replied:
Updated to "1.4". Though as an example it doesn't have to pass the check :)

caesarxuchao commented:
Thanks @liggitt. Tried the following and the rewrite still doesn't work:
```
echo "-X ${KUBE_GO_PACKAGE}/pkg/version.${key}=${val} -X k8s.io/client-go/pkg/version.${key}=${val} -X vendor/k8s.io/client-go/pkg/version.${key}=${val} -X staging/src/k8s.io/client-go/pkg/version.${key}=${val}"
```

liggitt commented:
> Thanks @liggitt. Tried the following and the rewrite still doesn't work:
> 
> echo "-X ${KUBE_GO_PACKAGE}/pkg/version.${key}=${val} -X k8s.io/client-go/pkg/version.${key}=${val} -X vendor/k8s.io/client-go/pkg/version.${key}=${val} -X staging/src/k8s.io/client-go/pkg/version.${key}=${val}"

The go path is k8s.io/kubernetes/vendor/k8s.io/client-go/...

building from HEAD with this:
```
  echo "-X ${KUBE_GO_PACKAGE}/pkg/version.${key}=${val}"
  echo "-X ${KUBE_GO_PACKAGE}/vendor/k8s.io/client-go/pkg/version.${key}=${val}"
```
produces
```
User-Agent: kubectl/v1.7.0 (darwin/amd64) kubernetes/4c4c9e4
```

liggitt commented:
opened https://github.com/kubernetes/kubernetes/pull/42623 with the build fix

liggitt commented:
if we get https://github.com/kubernetes/kubernetes/pull/42623 merged, I don't feel strongly about patching this version check

caesarxuchao commented:
@liggitt may i get the lgtm? Thanks.

liggitt commented:
in the abstract, the bounded version check LGTM, given the v0.0.0 issue is a bug we can fix, I'm unsure whether it's worth updating the 1.5 release for that

caesarxuchao commented:
Hmm, let's still merge this. It's a pure improvement. We might break kubectl again, this pull will save people from debugging the issue again.

liggitt commented:
did the previous unbounded check allow kubectl 1.3 to work against 1.5?

caesarxuchao commented:
1.5 apiserver will hide the `apps` and the `policy` groups from 1.3 kubectl, so 1.3 `kubectl apiversions` won't  see those two groups. Other than this, 1.3 kubectl works.

liggitt commented:
so tightening it breaks kubectl 1.3, which is outside our 1 version skew support statement, but makes this not a pure improvement. I could go either way.

caesarxuchao commented:
Excuse my English, I used the wrong tense. I mean without the tightening, "1.5 apiserver is hiding the apps and the policy groups from 1.3 kubectl, so 1.3 kubectl apiversions doesn't see those two groups." With the tightening, 1.3 kubectl can see these two groups enabled, though it still cannot do any CRUD operations because the lack of the scheme. So this pull is a pure improvement.

liggitt commented:
ok, so 1.4 was the only version that had the "already registered" problem?

caesarxuchao commented:
Sorry,1.3 is affected by the "already registered" change, 1.2 is not. We should tighten the version to >=1.3 and <1.5.

liggitt commented:
/lgtm

caesarxuchao commented:
@mwielgus I think you are managing the 1.5 branch, could you approve this pr to the 1.5 branch?
